gui/layers/_vectors_layer/model.py

```diff
@@ -0,0 +1,615 @@
+
+from vispy.util.event import Event
+
+from .view import QtVectorsLayer
+from .view import QtVectorsControls
+
+
+@add_to_viewer
+class Vectors(Layer):
+    """
+    Vectors layer renders lines onto the image.
+    There are currently two data modes supported:
+        1) image-like vectors where every pixel defines its own vector
+        2) coordinate-like vectors where x-y position is not fixed to a grid
+    Supports ONLY 2d vector field
+
+    Properties
+    ----------
+    vectors : np.ndarray
+        array of shape (N,4) or (N, M , 2)
+
+    data : np.ndarray
+        ABC implementation.
+        returns vectors property setter/getter
+
+    _original_data : np.ndarray
+        Used by averaging and length adjustments.
+        Is set during layer construction.
+        Is NOT updated if data is adjusted by assigning the 'vectors' property
+
+    averaging : str
+        kernel over which to average from one of _kernel_dict
+        averaging.setter adjusts the underlying data
+        subscribe an observer by registering it with "averaging_bind_to"
+
+    width : int
+        width of the line in pixels
+
+    length : int or float
+        length of the line
+        length.setter adjusts the underlying data
+        subscribe an observer by registering it with "length_bind_to"
+
+    color : str
+        one of "get_color_names" from vispy.color
+
+    conector : str or np.ndarray
+        way to render line between coordinates
+        two built in types: "segment" or "strip"
+        third type is np.ndarray
+
+    mode : str
+        control panel mode
+
+    Attributes
+    ----------
+    Private attributes
+    -------
+    _kernel_dict
+    _kernel
+    _data_types
+    _data_type
+    _width
+    _color
+    _colors
+    _connector_types
+    _connector
+    _avg_dims
+    _averaging
+    _length
+    _avg_observers
+    _len_observers
+    _need_display_update
+    _need_visual_update
+    _raw_dat
+    _original_data
+    _current_shape
+    _vectors
+    _mode
+    _mode_history
+
+    Public attributes
+    -------
+    name
+
+
+    See vispy's line visual docs for more details:
+    http://api.vispy.org/en/latest/visuals.html#vispy.visuals.LineVisual
+    http://vispy.org/visuals.html
+    """
+
+    def __init__(
+        self, vectors,
+            width=1,
+            color='red',
+            connector='segments',
+            averaging='1x1',
+            length=1):
+
+        visual = LinesNode()
+        super().__init__(visual)
+
+        # map averaging type to tuple
+        self._kernel_dict = {'1x1': (1, 1),
+                             '3x3': (3, 3),
+                             '5x5': (5, 5),
+                             '7x7': (7, 7),
+                             '9x9': (9, 9),
+                             '11x11': (11, 11)}
+        self._kernel = self._kernel_dict['1x1']
+
+        # Store underlying data model
+        self._data_types = ('matrix', 'projection')
```

Comment:
Why do you call the data types `matrix` and `projection` here, but then in the top of the docs and the examples call them `image-like` and `coordinate-like`. I'd prefer consistency and lean towards `image` and `coords`.